File: cpp/79.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  bool dfs(const vector<vector<char> > &board, const string &word,
           int i, int j, int cur, vector<vector<bool> > &saw){
    if(cur == word.length())
      return true;
    if(i <0 || j < 0 || i == board.size() || j == board[0].size() ||
       saw[i][j] || word[cur] != board[i][j])
      return false;
    saw[i][j] = true;
    bool result = (dfs(board, word, i+1, j, cur+1, saw) ||
                   dfs(board, word, i-1, j, cur+1, saw) ||
                   dfs(board, word, i, j+1, cur+1, saw) ||
                   dfs(board, word, i, j-1, cur+1, saw));
    saw[i][j] = false;
    return result;
  }
  bool exist(vector<vector<char> > &board, string word) {
    int m = board.size();
    int n = board[0].size();
    vector<vector<bool> > saw;
    for(int i=0; i<m; i++)
      saw.push_back(vector<bool>(n, false));
    for(int i=0; i<m; i++)
      for(int j=0; j<n; j++)
        if(board[i][j] == word[0])
          if(dfs(board, word, i, j, 0, saw))
            return true;
    return false;
  }
};


int main(){
  auto sol = Solution();
}

File: python/79.py
class Solution(object):
    @staticmethod
    def search(board, word, idx, i, j, vit):
        if idx == len(word):
            return True
        if i < 0 or i >= len(board) or\
           j < 0 or j >= len(board[0]) or\
           board[i][j] != word[idx]:
            return False
        vit.add((i, j))
        for di, dj in ((1, 0), (0, 1), (-1, 0), (0, -1)):
            ni, nj = i+di, j+dj
            if (ni, nj) not in vit:
                if Solution.search(board, word, idx+1, ni, nj, vit):
                    return True
        vit.remove((i, j))
        return False

    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        n = len(board)
        if not n:
            return False
        m = len(board[0])
        if not m:
            return False
        vit = set()
        for i, j in itertools.product(range(n), range(m)):
            if board[i][j] == word[0]:
                if Solution.search(board, word, 0, i, j, vit):
                    return True
        return False
